core/src/main/java/io/parsingdata/metal/token/Until.java

```diff
@@ -114,12 +116,17 @@ private Trampoline<Optional<ParseState>> iterate(final Environment environment,
 
     private Trampoline<Optional<ParseState>> parseSlice(final Environment environment, final BigInteger currentSize, final BigInteger stepSize, final BigInteger maxSize, final Slice slice) {
         return (currentSize.compareTo(ZERO) == 0 ? Optional.of(environment.parseState) : environment.parseState.add(new ParseValue(name, this, slice, environment.encoding)).seek(environment.parseState.offset.add(currentSize)))
-            .map(preparedParseState -> terminator.parse(environment.withParseState(preparedParseState)))
+            .map(preparedParseState -> parseTerminator(environment, preparedParseState))
             .orElseGet(Util::failure)
-            .map(parsedParseState -> complete(() -> success(parsedParseState)))
+            .map(parseState -> complete(() -> success(parseState)))
             .orElseGet(() -> intermediate(() -> iterate(environment, currentSize.add(stepSize), stepSize, maxSize)));
     }
 
+    private Optional<ParseState> parseTerminator(final Environment environment, final ParseState parseStateExcludingTerminator) {
+        return terminator.parse(environment.withParseState(parseStateExcludingTerminator))
+            .map(parseStateIncludingTerminator -> includeTerminator ? parseStateIncludingTerminator : parseStateExcludingTerminator);
+    }
```

Comment:
We decide to move the logic to a separate method, so we can give the different parseStates meaningful names and without cluttering the method above. But if that is not the case, then we can inline this method ofcourse.